### PR TITLE
MULE-18057: Cursors are not properly reset in NTLM dancing

### PR DESCRIPTION
**Pull request by fsgonz**

### CODE REVIEW

File: src/test/java/org/mule/service/http/impl/functional/client/RequestStreamingNtlmHttpClientPostStreamingTestCase.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+import org.junit.After;
+import org.junit.Before;
+
+import io.qameta.allure.Story;
+import io.qameta.allure.junit4.DisplayName;
+
+@Story(STREAMING)
+@DisplayName("Validates that the POST cursor body is preserved on NTLM authentication with streaming on request")
+public class RequestStreamingNtlmHttpClientPostStreamingTestCase extends CursorNtlmHttpClientPostStreamingTestCase {
```
Comment (eze210) [inline src/test/java/org/mule/service/http/impl/functional/client/RequestStreamingNtlmHttpClientPostStreamingTestCase.java]: Is it ok to extend CursorNtlmHttpClientPostStreamingTestCase? The other case does the same

Comment (eze210) [reply src/test/java/org/mule/service/http/impl/functional/client/RequestStreamingNtlmHttpClientPostStreamingTestCase.java]: It seems that you wanted to extend AbstractNtlmHttpClientPostStreamingTestCase instead

File: src/test/java/org/mule/service/http/impl/functional/client/AbstractNtlmHttpClientPostStreamingTestCase.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    if (authorization == null) {
+      response.statusCode(UNAUTHORIZED.getStatusCode());
+      response.addHeader("WWW-Authenticate", "NTLM");
+
+    } else if (authorization.equals("NTLM TlRMTVNTUAABAAAAAYIIogAAAAAoAAAAAAAAACgAAAAFASgKAAAADw==")) {
+      response.statusCode(UNAUTHORIZED.getStatusCode());
+      response.addHeader("WWW-Authenticate", "NTLM TlRMTVNTUAACAAAAAAAAACgAAAABggAAU3J2Tm9uY2UAAAAAAAAAAA==");
+
+    } else if (authorization
+        .equals("NTLM TlRMTVNTUAADAAAAGAAYAEgAAAAYABgAYAAAABQAFAB4AAAADAAMAIwAAAAaABoAmAAAAAAAAACyAAAAAYIAAgUBKAoAAAAPrYfKbe/jRoW5xDxHeoxC1gBmfWiS5+iX4OAN4xBKG/IFPwfH3agtPEia6YnhsADTVQBSAFMAQQAtAE0ASQBOAE8AUgBaAGEAcABoAG8AZABCAEEALQBGAEcATwBOAFoAQQAtAE8AUwBYAA==")) {
```
Comment (eze210) [inline src/test/java/org/mule/service/http/impl/functional/client/AbstractNtlmHttpClientPostStreamingTestCase.java]: Extract this and the other String constants

File: src/main/java/org/mule/service/http/impl/service/client/CursorNonBlockingInputStreamFeeder.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.ning.http.client.providers.grizzly.FeedableBodyGenerator;
+import com.ning.http.client.providers.grizzly.NonBlockingInputStreamFeeder;
+
+/**
+ * Inputstream feeder used so that we guarantee that a cursor inputstream is properly reset.
+ * 
+ * @since 4.3.0
```
Comment (eze210) [inline src/main/java/org/mule/service/http/impl/service/client/CursorNonBlockingInputStreamFeeder.java]: Should we specify runtime version or service version?

File: src/main/java/org/mule/service/http/impl/service/client/CursorInputStreamBodyGenerator.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+public class CursorInputStreamBodyGenerator extends InputStreamBodyGenerator {
+
+  public CursorInputStreamBodyGenerator(CursorStream inputStream) {
+    super(inputStream);
+  }
+
+  @Override
+  public Body createBody() throws IOException {
+    if (!inputStream.markSupported()) {
+      ((CursorStream) inputStream).seek(0);
```
Comment (eze210) [inline src/main/java/org/mule/service/http/impl/service/client/CursorInputStreamBodyGenerator.java]: Why doesn't it catch the seek(0) exception as in CursorNonBlockingInputStreamFeeder#reset() ?

File: src/main/java/org/mule/service/http/impl/service/client/CursorInputStreamBodyGenerator.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+import org.mule.runtime.api.streaming.bytes.CursorStream;
+
+import com.ning.http.client.Body;
+import com.ning.http.client.generators.InputStreamBodyGenerator;
+
+/**
+ * Input Stream Body Generator which properly resets input streams
+ * 
+ * @since 4.3.0
```
Comment (eze210) [inline src/main/java/org/mule/service/http/impl/service/client/CursorInputStreamBodyGenerator.java]: runtime or service?

File: src/test/java/org/mule/service/http/impl/functional/client/RequestStreamingNtlmHttpClientPostStreamingTestCase.java (hunk 1 above)
Comment (fsgonz) [reply src/test/java/org/mule/service/http/impl/functional/client/RequestStreamingNtlmHttpClientPostStreamingTestCase.java]: In one case I had to extend NtlmHttpClientPostStreamingTestCase. The idea is to test this without requestStreaming.

File: src/main/java/org/mule/service/http/impl/service/client/CursorNonBlockingInputStreamFeeder.java (hunk 3 above)
Comment (fsgonz) [reply src/main/java/org/mule/service/http/impl/service/client/CursorNonBlockingInputStreamFeeder.java]: Service version. It's correct your comment.

File: src/main/java/org/mule/service/http/impl/service/client/CursorInputStreamBodyGenerator.java (hunk 5 above)
Comment (fsgonz) [reply src/main/java/org/mule/service/http/impl/service/client/CursorInputStreamBodyGenerator.java]: service

**Comment by fsgonz** [comment]
--validate

**Pull request merged**